src/java_bytecode/java_string_library_preprocess.cpp

```diff
@@ -1169,6 +1174,302 @@ codet java_string_library_preprocesst::make_string_to_char_array_code(
+  pointer_typet pointer_of_pointer;
+  pointer_of_pointer.copy_to_subtypes(empty_pointer);
+  member_exprt data_member(deref_objs, "data", pointer_of_pointer);
+  plus_exprt data_pointer_plus_index(
+    data_member, from_integer(index, java_int_type()), data_member.type());
+  dereference_exprt data_at_index(
+    data_pointer_plus_index, data_pointer_plus_index.type().subtype());
+  return data_at_index;
+}
+
+/// Helper for format function. Adds code:
+/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// Object* arg_i=get_object_at_index(argv)
+/// string_expr arg_i_string_expr
+/// int arg_i_int
+/// ....
+/// if(arg_i!=NULL)
+///   if arg_i.@class_identifier=="java::java.lang.String"
+///     arg_i_string_expr = (string_expr)((String*)arg_i_as_string)
+///   arg_i_int=arg_i->value
+/// arg_i_struct = { string_expr=arg_i_string_expr;
+///   integer=arg_i_int; ...}
+/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// and returns `arg_i_struct`.
+///
+/// TODO: date_time and hash code are not implemented
+/// \param argv: reference to an array of references
+/// \param index: index of the desired argument
+/// \param structured_type: type for arguments of the internal format function
+/// \param loc: a location in the source
+/// \param symbol_table: the symbol table
+/// \param code: code block to which we are adding some assignements
+/// \return An expression of type `structured_type` representing the possible
+/// values of the argument at position `index` of `argv`.
+exprt java_string_library_preprocesst::make_argument_for_format(
+  const exprt &argv,
+  int index,
+  const struct_typet &structured_type,
+  const source_locationt &loc,
+  symbol_tablet &symbol_table,
+  code_blockt &code)
+{
+  // Declarations of the fields of arg_i_struct
+  // arg_i_struct is { arg_i_string_expr, tmp_int, tmp_char, ... }
+  struct_exprt arg_i_struct(structured_type);
+  std::list<exprt> field_exprs;
+  for(const auto & comp : structured_type.components())
+  {
+    const irep_idt &name=comp.get_name();
+    const typet &type=comp.type();
+    exprt field_expr;
+    if(name=="string_expr")
+    {
+      string_exprt arg_i_string_expr=fresh_string_expr(loc, symbol_table, code);
+      field_expr=arg_i_string_expr;
+    }
+    else
+    {
+      std::string tmp_name="tmp_"+id2string(name);
+      symbolt field_symbol=get_fresh_aux_symbol(
+            type, tmp_name, tmp_name, loc, ID_java, symbol_table);
+      field_expr=field_symbol.symbol_expr();
+      code.add(code_declt(field_expr));
+    }
+    field_exprs.push_back(field_expr);
+    arg_i_struct.copy_to_operands(field_expr);
+  }
+
+  // arg_i = argv[index]
+  exprt obj=get_object_at_index(argv, index);
+  symbolt object_symbol=get_fresh_aux_symbol(
+    obj.type(), "tmp_object", "tmp_object", loc, ID_java, symbol_table);
+  symbol_exprt arg_i=object_symbol.symbol_expr();
+  (void) allocate_dynamic_object_with_decl(
```

Comment:
What is happening with the return value? Is there something you would like to check?